### ISSUE TITLE
PR#122 makes test-charpoly fail to build

### ISSUE THREAD

**Issue by bgrenet** [linbox-team/linbox#124]
With the merge of #122, I got the following compilation error that I do not have without the merge (for `make test-charpoly`):

```
g++ -DHAVE_CONFIG_H -I. -I..  -DLinBoxTestOnly -O2 -Wall -g -DNDEBUG -U_LB_DEBUG -g -DNDEBUG -U_LB_DEBUG -I.. -fopenmp -fabi-version=6 -I/usr/local/include        -msse -msse2 -msse3 -mssse3 -msse4.1 -msse4.2 -mavx -MT test-charpoly.o -MD -MP -MF .deps/test-charpoly.Tpo -c -o test-charpoly.o test-charpoly.C
In file included from ../linbox/algorithms/bbcharpoly.h:44:0,
                 from ../linbox/solutions/charpoly.h:34,
                 from test-charpoly.C:46:
../linbox/ring/polynomial-ring.h:211:5: error: template-id ‘factor<template<class _Tp, class _Alloc> class std::vector>’ for ‘std::vector<LinBox::DensePolynomial<Givaro::ZRing<Givaro::Integer> > >& LinBox::PolynomialRing<Givaro::ZRing<Givaro::Integer>, Givaro::Dense>::factor(std::vector<LinBox::DensePolynomial<Givaro::ZRing<Givaro::Integer> > >&, std::vector<long unsigned int>&, const LinBox::DensePolynomial<Givaro::ZRing<Givaro::Integer> >&)’ does not match any template declaration
     LinBox::PolynomialRing<Givaro::ZRing<Givaro::Integer>,Givaro::Dense>::
     ^~~~~~
In file included from ../linbox/algorithms/bbcharpoly.h:44:0,
                 from ../linbox/solutions/charpoly.h:34,
                 from test-charpoly.C:46:
../linbox/ring/polynomial-ring.h:156:16: note: candidate is: template<template<class, class> class Vector, template<class> class Alloc, class B, typename std::enable_if<(std::is_same<LinBox::RingCategories::IntegerTag, LinBox::RingCategories::ModularTag>::value && std::is_same<B, Givaro::ZRing<Givaro::Integer> >::value), int>::type* <anonymous> > size_t LinBox::PolynomialRing<BaseRing, StorageTag>::factor(Vector<LinBox::DensePolynomial<BaseRing>, Alloc<LinBox::DensePolynomial<BaseRing> > >&, std::vector<long unsigned int>&, const Polynomial&) [with Vector = Vector; Alloc = Alloc; B = B; typename std::enable_if<(typename LinBox::FieldTraits<Field>::is_modular:: value && std::is_same<B, BaseRing>::value), int>::type* <anonymous> = <enumerator>; BaseRing = Givaro::ZRing<Givaro::Integer>; StorageTag = Givaro::Dense]
         size_t factor (Vector<Polynomial,Alloc<Polynomial> >& factors,
                ^~~~~~
In file included from ../linbox/algorithms/bbcharpoly.h:44:0,
                 from ../linbox/solutions/charpoly.h:34,
                 from test-charpoly.C:46:
../linbox/ring/polynomial-ring.h:281:5: error: template-id ‘factor<>’ for ‘std::vector<LinBox::DensePolynomial<LinBox::NTL_ZZ> >& LinBox::PolynomialRing<LinBox::NTL_ZZ, Givaro::Dense>::factor(std::vector<LinBox::DensePolynomial<LinBox::NTL_ZZ> >&, std::vector<long unsigned int>&, const LinBox::DensePolynomial<LinBox::NTL_ZZ>&)’ does not match any template declaration
     PolynomialRing<NTL_ZZ,Givaro::Dense>::factor (std::vector<DensePolynomial<NTL_ZZ> >& factors,
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../linbox/algorithms/bbcharpoly.h:44:0,
                 from ../linbox/solutions/charpoly.h:34,
                 from test-charpoly.C:46:
../linbox/ring/polynomial-ring.h:156:16: note: candidate is: template<template<class, class> class Vector, template<class> class Alloc, class B, typename std::enable_if<(std::is_same<LinBox::RingCategories::IntegerTag, LinBox::RingCategories::ModularTag>::value && std::is_same<B, LinBox::NTL_ZZ>::value), int>::type* <anonymous> > size_t LinBox::PolynomialRing<BaseRing, StorageTag>::factor(Vector<LinBox::DensePolynomial<BaseRing>, Alloc<LinBox::DensePolynomial<BaseRing> > >&, std::vector<long unsigned int>&, const Polynomial&) [with Vector = Vector; Alloc = Alloc; B = B; typename std::enable_if<(typename LinBox::FieldTraits<Field>::is_modular:: value && std::is_same<B, BaseRing>::value), int>::type* <anonymous> = <enumerator>; BaseRing = LinBox::NTL_ZZ; StorageTag = Givaro::Dense]
         size_t factor (Vector<Polynomial,Alloc<Polynomial> >& factors,
                ^~~~~~
Makefile:2438: recipe for target 'test-charpoly.o' failed
make: *** [test-charpoly.o] Error 1
```

**Comment by jgdumas** [comment]
Indeed PR122 does not address factor when NTL is enabled ...

**Comment by ClementPernet** [comment]
Fixed in #127
